### PR TITLE
Improvements in BO/BOLFI

### PR DESCRIPTION
**Pull request by vuolleko**
Improvements in BO/BOLFI:
- use seed
- use priors (if available)
- using the priors requires gradient information, which can be given by the user, or if not, is evaluated numerically
- replace previously used optimization (based on differential evolution) to one based on multiple runs of L-BFGS-B, used in acquisition of points and e.g. MAP estimate

### CODE REVIEW

File: elfi/methods/bo/acquisition.py [hunk 1]
```diff
@@ -109,30 +136,54 @@ class LCBSC(AcquisitionBase):
...
+        Parameters
+        ----------
+        x : numpy.array
+        t : int
+            Current iteration (starting from 0).
+        """
         if not isinstance(t, int):
             raise ValueError("Parameter 't' should be an integer.")
 
-        mean, var = self.model.predict(x, noiseless=True)
```
Comment (lintusj1) [inline elfi/methods/bo/acquisition.py]: Could you clarify why not to use the noiseless prediction? The BOLFI paper has this form.

File: elfi/methods/bo/acquisition.py [hunk 2]
```diff
@@ -109,30 +136,54 @@ class LCBSC(AcquisitionBase):
...
-        return mean - np.sqrt(self.beta(t) * var)
+        mean, var = self.model.predict(x)
+        return mean - np.sqrt(self._beta(t) * var)
+
+    def evaluate_grad(self, x, t=None):
```
Comment (lintusj1) [inline elfi/methods/bo/acquisition.py]: at some places `_grad` is used and in others `_gradient`. We should decide which form.

File: elfi/methods/bo/gpy_regression.py [hunk 3]
```diff
@@ -178,6 +178,11 @@ def predictive_gradients(self, x):
 
         return self._gp.predictive_gradients(x)
 
+    def predictive_gradient_mean(self, x):
```
Comment (lintusj1) [inline elfi/methods/bo/gpy_regression.py]: `_grad` or `_gradient`

File: elfi/methods/bo/utils.py [hunk 4]
```diff
@@ -53,9 +55,93 @@ def sum_of_rbf_kernels(point, kern_centers, kern_ampl, kern_scale):
     return ret
 
 
-def stochastic_optimization(fun, bounds, maxiter=1000, polish=True):
+def stochastic_optimization(fun, bounds, maxiter=1000, polish=True, seed=0):
```
Comment (lintusj1) [inline elfi/methods/bo/utils.py]: Could we remove this? It is used only in `BayesianOptimization.extract_result` but that could be easily replaced with the minimize function or then just directly use this `differential_evolution` there

File: elfi/methods/bo/utils.py [hunk 5]
```diff
@@ -53,9 +55,93 @@ def sum_of_rbf_kernels(point, kern_centers, kern_ampl, kern_scale):
...
+        result = fmin_l_bfgs_b(fun, inits[ii, :], fprime=grad, bounds=bounds, maxiter=maxiter)
+        locs.append(result[0])
+        vals[ii] = result[1]
+
+    # Return the optimal case
+    ind_min = np.argmin(vals)
+    return locs[ind_min], vals[ind_min]
+
+
+def numerical_grad_logpdf(x, *params, distribution=None, **kwargs):
```
Comment (lintusj1) [inline elfi/methods/bo/utils.py]: `grad` or `gradient`

File: elfi/methods/bo/utils.py [hunk 6]
```diff
@@ -53,9 +55,93 @@ def sum_of_rbf_kernels(point, kern_centers, kern_ampl, kern_scale):
...
+        parameters of the model
+    distribution : ScipyLikeDistribution or a distribution from Scipy
+
+    Returns
+    -------
+    grad_logpdf : ndarray
+       Gradient of the log of the probability density function evaluated at x
+    """
+
+    # due to the common scenario of logpdf(x) = -inf, multiple confusing warnings could be generated
```
Comment (lintusj1) [inline elfi/methods/bo/utils.py]: I'd like to know where do these warnings come from, the `logpdf` or `numdifftools.Grafient` method?

File: elfi/methods/methods.py [hunk 7]
```diff
@@ -788,7 +791,8 @@ def __init__(self, model, target=None, outputs=None, batch_size=1,
         if initial_evidence % self.batch_size != 0:
             raise ValueError('Initial evidence must be divisible by the batch size')
 
-        self.acquisition_method = acquisition_method or LCBSC(target_model)
+        priors = [self.model[p] for p in self.parameters]
```
Comment (lintusj1) [inline elfi/methods/methods.py]: We could make a utility function that constructs a prior `rvs` function from the model accepting also the seed (or even a distribution object). It would be usable elsewhere as well.

File: elfi/methods/methods.py [hunk 8]
```diff
@@ -788,7 +791,8 @@ def __init__(self, model, target=None, outputs=None, batch_size=1,
...
+        self.acquisition_method = acquisition_method or LCBSC(target_model, priors=priors)
```
Comment (lintusj1) [inline elfi/methods/methods.py]: This should be given the seed as well.

Comment (lintusj1) [inline elfi/methods/methods.py]: We could add the noise here too, it would be beneficial in the parallel execution.

File: elfi/methods/posteriors.py [hunk 9]
```diff
@@ -0,0 +1,288 @@
+import logging
```
Comment (lintusj1) [inline elfi/methods/posteriors.py]: I'd prefer that as long as there is just one class, this would be kept in the result module, but you can decide.

File: elfi/model/elfi_model.py [hunk 10]
```diff
@@ -165,7 +165,7 @@ def __init__(self, seed=None, batch_size=None, observed=None, pool=None):
 
         # Extract the seed from numpy RandomState. Alternative would be to use
         # os.urandom(4) casted as int.
-        self.seed = seed or np.random.RandomState().get_state()[1][0]
+        self.seed = np.random.RandomState().get_state()[1][0] if seed is None else seed
```
Comment (lintusj1) [inline elfi/model/elfi_model.py]: Why the change?

File: elfi/model/extensions.py [hunk 11]
```diff
@@ -16,6 +23,7 @@ def __init__(self, name=None):
         """
         self._name = name or self.__class__.__name__
 
+    @classmethod
```
Comment (lintusj1) [inline elfi/model/extensions.py]: Now this will be only usable with static classes. I would perhaps keep it as it was or make a separate definition for non static classes.

File: elfi/model/extensions.py [hunk 12]
```diff
@@ -50,23 +59,28 @@ def pdf(this, x, *params, **kwargs):
...
         Returns
         -------
-        pdf : ndarray
+        logpdf : ndarray
            Log of the probability density function evaluated at x
         """
-        raise NotImplementedError
+        p = this.pdf(x, *params, **kwargs)
+        if p == 0:
+            return -np.inf
```
Comment (lintusj1) [inline elfi/model/extensions.py]: I think this is not going to work with vectorized arguments. Also I think `np.log` returns -inf for zeros automatically. Some distributions may be more conveniently written directly as logpdf and then pdf is taken as the exponent.

File: README.md [hunk 13]
```diff
@@ -1,24 +1,96 @@
...
+ELFI depends on several other Python packages, which have their own dependencies. 
+Resolving these may sometimes go wrong:
+- If you receive an error about missing `numpy`, please install it first.
+- If you receive an error about `yaml.load`, install `pyyaml`.
+- On OS X with Anaconda virtual environment say `conda install python.app` and then use 
+`pythonw` instead of `python`.
+- Note that ELFI currently supports Python 3.5-3.6 only, although 3.x may work as well, 
+so try `pip3 install elfi`.
+
+Citing ELFI
```
Comment (vuolleko) [inline README.md]: ok

File: elfi/methods/bo/acquisition.py [hunk 14]
```diff
@@ -1,14 +1,14 @@
 import logging
-
 import numpy as np
+
```
Comment (vuolleko) [inline elfi/methods/bo/acquisition.py]: Yes, that's the PEP 8 convention, and I try to follow that as well. Fixed.

File: elfi/methods/bo/acquisition.py (hunk 1 above)
Comment (vuolleko) [inline elfi/methods/bo/acquisition.py]: Good catch.

File: elfi/methods/bo/acquisition.py (hunk 2 above)
Comment (vuolleko) [inline elfi/methods/bo/acquisition.py]: Ok, I'll change all to grad, since even [Wikipedia](https://en.wikipedia.org/wiki/Gradient) says: "The notation grad f is also commonly used for the gradient."

File: elfi/methods/bo/utils.py (hunk 4 above)
Comment (vuolleko) [inline elfi/methods/bo/utils.py]: Let's not remove perfectly working stuff... :) It's in utils after all.

File: elfi/methods/bo/utils.py [hunk 15]
```diff
@@ -53,9 +55,93 @@ def sum_of_rbf_kernels(point, kern_centers, kern_ampl, kern_scale):
...
+    if priors[0] is None:
+        # Sample initial points uniformly within bounds
+        random_state = random_state or np.random.RandomState()
+        for ii in range(len(priors)):
+            inits[:, ii] = random_state.uniform(*bounds[ii], n_inits)
+
+    else:
+        # Sample priors for initialization points
+        prior_names = [p.name for p in priors]
+        inits_dict = priors[0].model.generate(n_inits, outputs=prior_names)
```
Comment (vuolleko) [inline elfi/methods/bo/utils.py]: Oh, then the documentation is a bit confusing: `ElfiModel.generate` says: "Generates a batch of outputs using the global seed."

File: elfi/methods/bo/utils.py (hunk 6 above)
Comment (vuolleko) [inline elfi/methods/bo/utils.py]: Well you can try. :) If there's a zero/inf, several warnings will be shown.

File: elfi/methods/methods.py (hunk 8 above)
Comment (vuolleko) [inline elfi/methods/methods.py]: Good, then BOLFI can directly use BO's `__init__`.

File: elfi/methods/posteriors.py (hunk 9 above)
Comment (vuolleko) [inline elfi/methods/posteriors.py]: I dislike long files. :)

File: elfi/model/elfi_model.py (hunk 10 above)
Comment (vuolleko) [inline elfi/model/elfi_model.py]: Didn't work with `seed=0`.

File: elfi/model/extensions.py (hunk 12 above)
Comment (vuolleko) [inline elfi/model/extensions.py]: Good point about vectors, I'll fix that.
np.log(0) does return -inf, but generates a warning.

File: elfi/model/extensions.py (hunk 11 above)
Comment (vuolleko) [inline elfi/model/extensions.py]: The MA2 example won't work with the old version (`logpdf` calling `this.pdf`).

We can add a definition for non-static classes once we need it. Of course, even better if there's a solution that works for both.

File: elfi/methods/methods.py (hunk 7 above)
Comment (vuolleko) [inline elfi/methods/methods.py]: That would be good, along with `pdf` etc. :)
TODO

File: elfi/methods/bo/utils.py [hunk 16]
```diff
@@ -53,9 +55,93 @@ def sum_of_rbf_kernels(point, kern_centers, kern_ampl, kern_scale):
...
+    
+    Parameters
+    ----------
+    fun : callable
+        Function to minimize.
+    grad : callable
+        Gradient of fun.
+    bounds : list of tuples
+        Bounds for each parameter.
+    priors : list of elfi.Priors, or list of Nones
```
Comment (vuolleko) [inline elfi/methods/bo/utils.py]: Agree. TODO

File: elfi/methods/bo/acquisition.py [hunk 17]
```diff
@@ -23,28 +23,53 @@ class AcquisitionBase:
                 bounds : tuple of length 'input_dim' of tuples (min, max)
             and methods
                 evaluate(x) : function that returns model (mean, var, std)
-    n_samples : None or int
-        Total number of samples to be sampled, used when part of an
-        AcquisitionSchedule object (None indicates no upper bound)
+    priors : list of elfi.Priors, optional
```
Comment (vuolleko) [inline elfi/methods/bo/acquisition.py]: Yes, that would be good. But maybe that can be a TODO for now.

File: elfi/model/extensions.py (hunk 12 above)
Comment (lintusj1) [inline elfi/model/extensions.py]: I'd keep this as not implemented because sometimes the `pdf` function may call this.

File: elfi/model/extensions.py (hunk 11 above)
Comment (lintusj1) [inline elfi/model/extensions.py]: So this is not meant to be used by others implementing their custom distributions?

File: elfi/model/extensions.py (hunk 12 above)
Comment (lintusj1) [inline elfi/model/extensions.py]: That way you can remove the @classmethod decorators. Implementing subclasses can add them if they need them.

File: elfi/model/extensions.py (hunk 11 above)
Comment (vuolleko) [inline elfi/model/extensions.py]: ? Isn't the MA2 example a good example of how they can use this?

I think we're comparing apples and oranges here? :)
